### PR TITLE
node(svm-watcher): Add type to validate message account data

### PR DESCRIPTION
**Pull request by johnsaigle**
- Add new MessageAcccountData type to validate message account data is the right size and has a valid prefix.
- Modify relevant function signatures to accept the new type instead of []byte.
- Modify the recent changes to the SVM watcher's unit test by using the new safe struct instead of manually building the message account data.

### CODE REVIEW

File: node/pkg/watchers/solana/client.go [hunk 1]
```diff
@@ -1122,10 +1171,15 @@ func (s *SolanaWatcher) getLatestFinalizedBlockNumber() uint64 {
...
+	if messageAccountData == nil {
+		return nil, errors.New("message account data is nil")
+	}
+
 	prop := &MessagePublicationAccount{}
-	// Skip the b"msg" prefix
-	if err := borsh.Deserialize(prop, data[3:]); err != nil {
+	// Skip the prefix and deserialize the rest of the data.
+	// SECURITY: The MessageAccountData type ensures that the data is at least 3 bytes long.
+	if err := borsh.Deserialize(prop, messageAccountData.data[MessageAccountDataMinLength:]); err != nil {
```
Comment (djb15) [inline node/pkg/watchers/solana/client.go]: I'm wondering whether we should stick with `3` or some other specified constant of value 3 here instead of `MessageAccountDataMinLength`. As the docstring for the variable declaration suggests, it's possible that we could increase `MessageAccountDataMinLength` to be more restrictive, but here we already want to just skip the first 3 bytes

File: node/pkg/watchers/solana/client_test.go [hunk 2]
```diff
@@ -107,3 +109,116 @@ func TestIsPossibleWormholeMessageFail(t *testing.T) {
...
+		},
+		{
+			name:               "failure -- nil argument",
+			messageAccountData: nil,
+			want:               &MessagePublicationAccount{},
+			wantErr:            true,
+		},
+		{
+			name:               "failure -- no data following prefix (msg)",
+			messageAccountData: &MessageAccountData{[]byte("msg")},
```
Comment (djb15) [inline node/pkg/watchers/solana/client_test.go]: What are the errors we're expecting? Can you expand the test to check for the specific error strings?

The new checks are for `length < 3` which these aren't, so it must be failing somewhere else in the existing logic?

File: node/pkg/watchers/solana/client.go (hunk 1 above)
Comment (johnsaigle) [reply node/pkg/watchers/solana/client.go]: Good point. In this case it makes sense to decouple the constants. I'll make another one that's called "PrefixLength"

File: node/pkg/watchers/solana/client_test.go (hunk 2 above)
Comment (johnsaigle) [reply node/pkg/watchers/solana/client_test.go]: Borsh fails to deserialize the data into the struct and give an error message to that effect. I could include that error string here if you want, though we'll need to update the tests if the `borsh-go` library changes.

Comment (djb15) [reply node/pkg/watchers/solana/client_test.go]: No that's fine, maybe also just add another test case with <3 bytes but not nil

Comment (johnsaigle) [reply node/pkg/watchers/solana/client_test.go]: I added detection for the error strings before you replied 😅 . I also added the unit test case you described.

**Review by djb15 — DISMISSED** [review]

File: node/pkg/watchers/solana/client.go [hunk 3]
```diff
@@ -205,6 +244,26 @@ const (
...
+// NewMessageAccountData creates a new MessageAccountData from the given data.
+// SECURITY:
+// 1. Data must be non-empty.
+// 2. Data must begin with b"msg" for regular message accounts or b"msu" for unreliable accounts (used by the shim contract).
+// The program must never parse account data as a message account if it is does not meet these requirements.
+func NewMessageAccountData(data []byte) *MessageAccountData {
+	if len(data) < MessageAccountDataMinLength {
+		return nil
+	}
+	if !bytes.HasPrefix(data, []byte(accountPrefixReliable)) && !bytes.HasPrefix(data, []byte(accountPrefixUnreliable)) {
```
Comment (mdulin2) [inline node/pkg/watchers/solana/client.go:258]: I like the explicit parsing of the prefixes.

File: node/pkg/watchers/solana/client.go [hunk 4]
```diff
@@ -968,38 +1029,42 @@ func (s *SolanaWatcher) processAccountSubscriptionData(_ context.Context, logger
...
 	}
 
-	data, err = base64.StdEncoding.DecodeString(value.Account.Data[0])
+	// SECURITY: Bounds check the Data slice to ensure it's not empty.
+	var accountDataBase64 string
+	if len(value.Account.Data) == 0 {
+		return nil
+	} else {
+		// The data is received as an array of base64 strings, but we only care about the first one.
+		accountDataBase64 = value.Account.Data[0]
```
Comment (mdulin2) [inline node/pkg/watchers/solana/client.go:1139]: Why does this have more than one element?

**Review by mdulin2 — DISMISSED** [review]

File: node/pkg/watchers/solana/client.go (hunk 4 above)
Comment (johnsaigle) [reply node/pkg/watchers/solana/client.go:1139]: We're parsing the JSON as a slice of strings: https://github.com/wormhole-foundation/wormhole/blob/601307f295513dcff35de20446bad02029443664/node/pkg/watchers/solana/client.go#L123

I'm not totally sure if that's correct. Maybe it's possible that it's really just a single string and the watcher is not representing it correctly.

**Review by djb15 — DISMISSED** [review]

File: node/pkg/watchers/solana/client.go [hunk 5]
```diff
@@ -156,6 +170,31 @@ const (
...
+	//   consistency_level      u8        =  1
+	//   vaa_time               u32       =  4
+	//   vaa_signature_account  Pubkey    = 32
+	//   submission_time        u32       =  4
+	//   nonce                  u32       =  4
+	//   sequence               u64       =  8
+	//   emitter_chain          u16       =  2
+	//   emitter_address        [u8; 32]  = 32
+	//   payload_len            u32       =  4  (Borsh Vec length prefix)
+	//   payload                [u8; N]   =  N  (variable, 0 minimum)
```
Comment (pleasew8t) [inline node/pkg/watchers/solana/client.go]: Wouldn't this more accurately be `[u8; payload_len]`?

File: node/pkg/watchers/solana/client.go (hunk 4 above)
Comment (pleasew8t) [reply node/pkg/watchers/solana/client.go:1139]: According to https://solana.com/docs/rpc/websocket/programsubscribe it looks like Data is a 2-element array containing:

`[ "encoded data", "encoding type" ]`

The example in the docs:

```
"data": [
    "11116bv5nS2h3y12kD1yUKeMZvGcKLSjQgX6BeV7u1FrjeJcKfsHPXHRDEHrBesJhZyqnnq9qJeUuF7WHxiuLuL5twc38w2TXNLxnDbjmuR",
    "base58"
],
```

**Review by pleasew8t — DISMISSED** [review]

**Review by djb15 — DISMISSED** [review]

**Review by djb15 — DISMISSED** [review]

**Comment by johnsaigle** [comment]
@djb15 @pleasew8t @mdulin2  could you please re-review? the only changes since your previous approvals are the ones in the unit test file

File: node/pkg/watchers/solana/client_test.go [hunk 6]
```diff
@@ -265,56 +269,64 @@ func TestParseMessagePublicationAccount(t *testing.T) {
...
+			errStr: errStringPrefix,
 		},
 		{
-			name:               "failure -- data too short",
-			messageAccountData: MessageAccountData{[]byte("ms")},
-			want:               &MessagePublicationAccount{},
-			errStr:             errStringPrefix,
+			name: "failure -- data too short",
+			messageAccountData: func(t *testing.T) MessageAccountData {
+				return MessageAccountData{[]byte("ms")}
```
Comment (djb15) [inline node/pkg/watchers/solana/client_test.go:287]: Why not use `mustNewMessageAccountData` here too?

Comment (johnsaigle) [reply node/pkg/watchers/solana/client_test.go:287]: I added some clarification in the latest commit

**Review by djb15 — DISMISSED** [review]

**Review by mdulin2 — APPROVED** [review]

**Review by djb15 — APPROVED** [review]
